ethereum/contracts/Broker.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2;
+
+import "./Verifier.sol";
+import "./Decoder.sol";
+import "./Application.sol";
+
+contract Broker {
+    using Decoder for bytes;
+
+    mapping(address => bool) public applications;
+    Verifier public verifier;
+
+    constructor(address verifierAddr) public {
+        verifier = new Verifier(verifierAddr);
+    }
+
+    /**
+     * @dev registers new applications
+     * @param _appID address _appID is the contract address to be registered as a supported application
+     */
+    function registerApp(address _appID)
+        public
+    {
+        require(!applications[_appID], "Application is already registered");
+
+        // 0x0 is the value returned for not-yet created accounts.
+        // 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 is returned for accounts without code.
+        bytes32 accountHash = 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470;
+        bytes32 codehash;
+        assembly { codehash := extcodehash(_appID) }
+        require(codehash != 0x0 && codehash != accountHash, "Application must be a contract deployed to this network");
+
+        // TODO: Consider checking that the application implements the abstract Application method 'submit' according
+        //       to EIP-165 (incomplete implementation: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md)
+
+        applications[_appID] = true;
+    }
+
+    /**
+     * @dev routes the message to the specified application ID after verifying the operator's signature
+     * @param _data address _data expected type: Message { AppID [32]byte, Payload []byte }
+     * @param _signature address _signature expected type: Signature []byte
+     */
+    function submit(bytes memory _data, bytes memory _signature)
```

Comment:
Also the `_signature` field in `submit` was just a hack/convenience to get M4 out the door. On the substrate side, the signature is part of the data/message bytes:

In rough pseudocode:
```
message {
  []byte payload
  []byte signature
}
```

We'll still need to think of a suitable encoding scheme for Message, be it RLP or Scale, etc.

Reply:
what is the signature here for?

Reply:
Its to mirror the substrate side, and implement signature-based verification of messages. For example, on the Substrate side, this what the relayer sends to substrate:
https://github.com/Snowfork/polkadot-ethereum/blob/main/parachain/primitives/ethereum/src/message.rs#L15

I know @denalimarsh has implemented a verifier contract though I haven't had a proper chance to review it yet.

Of course post-M5 we'll be using light-client based verification, so we'll end up throwing this signature code away.

Reply:
@vgeddes right but on the substrate side, is this SignedMessage used by the broker? or only by other modules?

Reply:
On the Substrate side, things are still bit messy and ad-hoc. While I've implemented the primitives to verify signed messages, the verifier module isn't actually using those primitives yet. Currently, after no-op dummy verification, the broker submits an opaque Message to an application. The application then decodes that opaque message to a SignedMessage. Fixing all that on my TODO list for M5.

I suppose the ideal approach on both chains would be the following sequence:

* Bridge receives `OpaqueMessage`
* Bridge forwards `OpaqueMessage` to Broker
* Broker forwards `OpaqueMessage` to Verifier
* Verifier decodes `OpaqueMessage` into a ` SignedMessage { data, signature }`.
* Verifier verifies message, strips off the signature, and submits `VerifiedMessage { data }` back to the broker
* Broker dispatches `VerifiedMessage { data }` to target application

Reply:
Removed signature-based verification in the Broker contract and replaced with chronological-based verification using block number and event hash in https://github.com/Snowfork/polkadot-ethereum/pull/75/commits/58850b4f6d0450ebd23b31d7deabe839608c37d2, https://github.com/Snowfork/polkadot-ethereum/pull/75/commits/f9121d56afc6b7403efaffe93c6a2d3b9a7813e5